src/backend/commands/matview.c

```diff
@@ -861,45 +861,43 @@ refresh_by_match_merge(Oid matviewOid, Oid tempOid, Oid relowner,
 	 * Create the temporary "diff" table.
 	 *
 	 * Temporarily switch out of the SECURITY_RESTRICTED_OPERATION context,
-	 * because you cannot create temp tables in SRO context.  For extra
-	 * paranoia, add the composite type column only after switching back to
-	 * SRO context.
+	 * because you cannot create temp tables in SRO context.  In GPDB use
+	 * "SELECT alias.*" rather than "alias.*::compositetype" in upstream to
+	 * support DISTRIBUTED BY clause.
 	 */
 	SetUserIdAndSecContext(relowner,
 						   save_sec_context | SECURITY_LOCAL_USERID_CHANGE);
-<<<<<<< HEAD
 
 	/* Get distribute key of matview */
 	distributed = TextDatumGetCString(DirectFunctionCall1(pg_get_table_distributedby,
 														   ObjectIdGetDatum(tempOid)));
 
-	/* Start building the query for creating the diff table. */
 	resetStringInfo(&querybuf);
 	appendStringInfo(&querybuf,
-					 "CREATE TEMP TABLE %s AS "
-					 "SELECT mv.ctid AS tid, mv.gp_segment_id as sid, newdata.* "
-=======
+					 "CREATE TEMP TABLE %s (LIKE %s)",
+					 diffname, tempname);
+	if (SPI_exec(querybuf.data, 0) != SPI_OK_UTILITY)
+		elog(ERROR, "SPI_exec failed: %s", querybuf.data);
 	resetStringInfo(&querybuf);
 	appendStringInfo(&querybuf,
-					 "CREATE TEMP TABLE %s (tid pg_catalog.tid)",
+					 "ALTER TABLE %s ADD COLUMN tid pg_catalog.tid",
 					 diffname);
 	if (SPI_exec(querybuf.data, 0) != SPI_OK_UTILITY)
 		elog(ERROR, "SPI_exec failed: %s", querybuf.data);
-	SetUserIdAndSecContext(relowner,
-						   save_sec_context | SECURITY_RESTRICTED_OPERATION);
 	resetStringInfo(&querybuf);
 	appendStringInfo(&querybuf,
-					 "ALTER TABLE %s ADD COLUMN newdata %s",
-					 diffname, tempname);
+					 "ALTER TABLE %s ADD COLUMN sid pg_catalog.int",
+					 diffname);
 	if (SPI_exec(querybuf.data, 0) != SPI_OK_UTILITY)
 		elog(ERROR, "SPI_exec failed: %s", querybuf.data);
+	SetUserIdAndSecContext(relowner,
+						   save_sec_context | SECURITY_RESTRICTED_OPERATION);
```

Comment:
Should we switch the context just after creating the temporary table as it is done in the original patch?

Reply:
moved